backend/plonk/bn254/solidity.go

```diff
@@ -422,15 +384,134 @@ contract PlonkVerifier {
+      mstore8(add(mPtr, 0x80), 0)
+      mstore8(add(mPtr, 0x81), lenInBytes)
+      mstore8(add(mPtr, 0x82), 0)
+
+      // "BSB22-Plonk" = [42, 53, 42, 32, 32, 2d, 50, 6c, 6f, 6e, 6b,]
+      mstore8(add(mPtr, 0x83), 0x42)
+      mstore8(add(mPtr, 0x84), 0x53)
+      mstore8(add(mPtr, 0x85), 0x42)
+      mstore8(add(mPtr, 0x86), 0x32)
+      mstore8(add(mPtr, 0x87), 0x32)
+      mstore8(add(mPtr, 0x88), 0x2d)
+      mstore8(add(mPtr, 0x89), 0x50)
+      mstore8(add(mPtr, 0x8a), 0x6c)
+      mstore8(add(mPtr, 0x8b), 0x6f)
+      mstore8(add(mPtr, 0x8c), 0x6e)
+      mstore8(add(mPtr, 0x8d), 0x6b)
+
+      // size domain
+      mstore8(add(mPtr, 0x8e), sizeDomain)
+
+      let success := staticcall(gas(), 0x2, mPtr, 0x8f, mPtr, 0x20)
+      if iszero(success) {
+        error_sha2_256()
+      }
+
+      let b0 := mload(mPtr)
+
+      // [b0         || one || dst || sizeDomain]
+      // <-64bytes ->  <-    1 byte each      ->
+      mstore8(add(mPtr, 0x20), one) // 1
+      
+      mstore8(add(mPtr, 0x21), 0x42) // dst
+      mstore8(add(mPtr, 0x22), 0x53)
+      mstore8(add(mPtr, 0x23), 0x42)
+      mstore8(add(mPtr, 0x24), 0x32)
+      mstore8(add(mPtr, 0x25), 0x32)
+      mstore8(add(mPtr, 0x26), 0x2d)
+      mstore8(add(mPtr, 0x27), 0x50)
+      mstore8(add(mPtr, 0x28), 0x6c)
+      mstore8(add(mPtr, 0x29), 0x6f)
+      mstore8(add(mPtr, 0x2a), 0x6e)
+      mstore8(add(mPtr, 0x2b), 0x6b)
+
+      mstore8(add(mPtr, 0x2c), sizeDomain) // size domain
+      success := staticcall(gas(), 0x2, mPtr, 0x2d, mPtr, 0x20)
+      if iszero(success) {
+        error_sha2_256()
+      }
+
+      // b1 is located at mPtr. We store b2 at add(mPtr, 0x20)
+
+      // [b0^b1      || two || dst || sizeDomain]
+      // <-64bytes ->  <-    1 byte each      ->
+      mstore(add(mPtr, 0x20), xor(mload(mPtr), b0))
+      mstore8(add(mPtr, 0x40), two)
+
+      mstore8(add(mPtr, 0x41), 0x42) // dst
+      mstore8(add(mPtr, 0x42), 0x53)
+      mstore8(add(mPtr, 0x43), 0x42)
+      mstore8(add(mPtr, 0x44), 0x32)
+      mstore8(add(mPtr, 0x45), 0x32)
+      mstore8(add(mPtr, 0x46), 0x2d)
+      mstore8(add(mPtr, 0x47), 0x50)
+      mstore8(add(mPtr, 0x48), 0x6c)
+      mstore8(add(mPtr, 0x49), 0x6f)
+      mstore8(add(mPtr, 0x4a), 0x6e)
+      mstore8(add(mPtr, 0x4b), 0x6b)
+
+      mstore8(add(mPtr, 0x4c), sizeDomain) // size domain
+
+      let offset := add(mPtr, 0x20)
+      success := staticcall(gas(), 0x2, offset, 0x2d, offset, 0x20)
+      if iszero(success) {
+        error_sha2_256()
+      }
+
+      // at this point we have mPtr = [ b1 || b2] where b1 is on 32byes and b2 in 16bytes.
+      // we interpret it as a big integer mod r in big endian (similar to regular decimal notation)
+      // the result is then 2**(8*16)*mPtr[32:] + mPtr[32:48]
+      res := mulmod(mload(mPtr), bb, r_mod) // <- res = 2**128 * mPtr[:32]
+      offset := add(mPtr, 0x10)
+      for {let i:=0} lt(i, 0x10) {i:=add(i,1)} // mPtr <- [xx, xx, ..,  | 0, 0, .. 0  ||    b2   ]
```

Comment:
Not sure about optimisations. Would unrolling the loop help? Otherwise there are implicit checks for overflow when doing add, no?

Reply:
I could try to unroll it. But I might get a 'stack too deep' error if the function is too big